wordcount/data.go

```diff
@@ -116,66 +116,115 @@ func splitData(fileName string, chunkSize int) (numMapFiles int, err error) {
+    bytes, err := file.ReadAt(chunk, fileOffset)
+    if err!=nil && err!=io.EOF {
         log.Fatal(err)
-        return 0, err
+        return chunk, 0
     }
-    
-    numMapFiles = 0
-    minOffset := 0
-    maxOffset := 0
-    lastNotWordCharOffset := 0
-    
-    // IMPORTANT: This code already supports non-ascii texts;
-    for currOffset, char:=range string(buffer[:]) {
-        isWordChar := unicode.IsLetter(char) || unicode.IsNumber(char) // Word chars must be letters or numeric digits;
-        charSize := utf8.RuneLen(char) // In bytes;
-        if currOffset-minOffset+charSize>chunkSize { // Current char cannot be included;
-            maxOffset = currOffset
-            if isWordChar {
-                maxOffset = lastNotWordCharOffset
-            }
-            if err = createFile(mapFileName(numMapFiles), buffer[minOffset:maxOffset]); err!=nil {
-                log.Fatal(err)
-                return numMapFiles, err
-            }
-            numMapFiles++
-            minOffset = maxOffset
-        } else if !isWordChar {
-            lastNotWordCharOffset = currOffset+charSize
+    if bytes<len(chunk) {
+        return chunk[:bytes], int64(bytes) // (*) "Trimming" chunk to bytesRead; this also implies this is a last chunk;
+    } else {
+        return chunk, int64(bytes)
+    }
+}
+
+func findChunkMaxOffset(chunk []byte, chunkSize int64) (chunkMaxOffset int64) {
+    chunkLength := int64(len(chunk))
+    if chunkLength<chunkSize { // Last chunk; see (*) above;
+        return chunkLength
+    }
+    offset := chunkSize-1
+    for { // Start searching...
+        char, charStartOffset, charBytes := findPreviousChar(chunk, offset, false)
+        if char==0 { // This should never happen;
+            return 0
+        } else if charStartOffset+charBytes<=chunkSize { // Check against chunkSize;
+            if unicode.IsLetter(char) || unicode.IsNumber(char) { // Word char;
+                nextChar, _ := utf8.DecodeRune(chunk[charStartOffset+charBytes:])
+                if unicode.IsLetter(nextChar) || unicode.IsNumber(nextChar) { // If nextChar is also a word char, found word split;
+                    offset = charStartOffset-1 // Update offset;
+                    prevNonWordChar, prevNonWordCharStartOffset, prevNonWordCharBytes := findPreviousChar(chunk, offset, true) // Search for the previous non-word char;
+                    if prevNonWordChar==0 { // This should never happen;
+                        return 0
+                    } else {
+                        return prevNonWordCharStartOffset + prevNonWordCharBytes;   
+                    }
+                } else { // Last char of a word;
+                    return charStartOffset + charBytes;
+                }
+            } else { // Non-word char;
+                return charStartOffset + charBytes;
+            }   
+        } else { // Char outside chunk size;
+            offset = charStartOffset-1 // Update offset;
```

Comment:
Paulo, a função findChunkMaxOffset(), como o nome indica, busca o offset máximo do chunk. Esse offset máximo é onde eu devo fazer a quebra do arquivo. Ele lê o chunk de trás pra frente, lendo cada caracter individualmente (isso é feito no loop infinito - comando for -  que chama a função findPreviousChar()). Para cada um dos caracteres lidos, a primeira coisa é verificar se ele não estoura chunkSize (linha 185). Se esse caracter estoura chunkSize, eu ignoro o caracter e simplesmente atualizo o offset do chunk (que é utilizado em findPreviousChar()) para continuar a busca retroativa de caracteres (linha 203). Se o caracter não estoura chunkSize, existem as seguintes situações:

1) O caracter não é letra nem número (linha 199): encontramos o offset máximo do chunk e podemos retornar;
2) O caracter é uma letra: nesse caso, existem 2 casos (linha 186):
       a) O caracter **seguinte** não é letra nem número (linha 196): o caracter corrente é a última letra ou número da palavra: encontramos o offset máximo do chunk e podemos retornar;
       b) O caracter seguinte **é letra ou número** (linha 188): o caracter corrente está no meio da palavra: devemos encontrar o offset do primeiro caracter não letra nem número que antecede o caracter corrente; isso é feito pela chamada da função findPreviousChar() (linha 190) com a flag nonWordCharOnly==true; o resultado dessa função é o que deve ser retornado nesse caso (linha 194);